### ISSUE TITLE
How the INNER_EOL_DOC_COMMENT element rendered?

### ISSUE THREAD

**Issue by izhangzhihao** [intellij-rust/intellij-rust#6100]
## Environment

```
CLion 2020.2.1
Build #CL-202.6948.80, built on August 25, 2020
Runtime version: 11.0.8+10-b944.31 amd64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
Linux 5.7.15-100.fc31.x86_64
GC: ParNew, ConcurrentMarkSweep
Memory: 1987M
Cores: 12
Registry: run.processes.with.pty=TRUE, documentation.show.toolbar=true, ide.allow.merge.buttons=false, search.everywhere.settings=true, debugger.watches.in.variables=false, cidr.max.intellisense.file.length=600000
Non-Bundled Plugins: name.kropp.intellij.makefile, org.jetbrains.plugins.localization, org.toml.lang, tanvd.grazi, org.rust.lang, rs.pest, izhangzhihao.rainbow.brackets
Current Desktop: KDE
```

## Problem description

Please see: https://github.com/izhangzhihao/intellij-rainbow-brackets/issues/483

## Steps to reproduce

Install rainbow brackets and open the `option` class, in line 59.

## Question

I want to know how the `INNER_EOL_DOC_COMMENT` rendered because I didn't see the internal structure.

**Comment by Undin** [comment]
We have two different level of highlighting inside doc comments:
- highlighting of basic markdown structures is done via lexer-based highlighter
- highlighting of rust code snippets is done via language injection, i.e. the plugin looks for code snippets and inject rust code into the corresponding line.

> I didn't see the internal structure.

Yep, currently, doc comments are leaf elements in PSI tree. But @vlad20012 is working on PSI to parse markdown inside Rust doc comments

**Comment by izhangzhihao** [comment]
What is the plan to finish this feature?

**Comment by Undin** [comment]
@izhangzhihao yep. Follow https://github.com/intellij-rust/intellij-rust/pull/6455

**Comment by izhangzhihao** [comment]
Well, it seems that the code :

```
//! ```
//! let optional = None;
//! check_optional(optional);
//!
//! let optional = Some(Box::new(9000));
//! check_optional(optional);
//!
//! fn check_optional(optional: Option<Box<i32>>) {
//!     match optional {
//!         Some(p) => println!("has value {}", p),
//!         None => println!("has no value"),
//!     }
//! }
//! ```
```

are shown as:

```
RsDocCodeFenceImpl(<DOC_CODE_FENCE>)
RsDocCodeFenceStartEndImpl(<DOC_CODE_FENCE_START_END>)
PsiWhiteSpace
PsiElement(<DOC_GAP>)
PsiWhiteSpace
PsiElement(<DOC_DATA>)
PsiWhiteSpace
PsiElement(<DOC_GAP>)
PsiWhiteSpace
PsiElement(<DOC_DATA>)
PsiWhiteSpace
PsiElement(<DOC_GAP>)
PsiWhiteSpace
PsiElement(<DOC_GAP>)
PsiWhiteSpace
PsiElement(<DOC_DATA>)
PsiWhiteSpace
PsiElement(<DOC_GAP>)
PsiWhiteSpace
PsiElement(<DOC_DATA>)
PsiWhiteSpace
PsiElement(<DOC_GAP>)
PsiWhiteSpace
PsiElement(<DOC_GAP>)
PsiWhiteSpace
PsiElement(<DOC_DATA>)
PsiWhiteSpace
PsiElement(<DOC_GAP>)
PsiWhiteSpace
PsiElement(<DOC_DATA>)
PsiWhiteSpace
PsiElement(<DOC_GAP>)
PsiWhiteSpace
PsiElement(<DOC_DATA>)
PsiWhiteSpace
PsiElement(<DOC_GAP>)
PsiWhiteSpace
PsiElement(<DOC_DATA>)
PsiWhiteSpace
PsiElement(<DOC_GAP>)
PsiWhiteSpace
PsiElement(<DOC_DATA>)
PsiWhiteSpace
PsiElement(<DOC_GAP>)
PsiWhiteSpace
PsiElement(<DOC_DATA>)
PsiWhiteSpace
PsiElement(<DOC_GAP>)
PsiWhiteSpace
RsDocCodeFenceStartEndImpl(<DOC_CODE_FENCE_START_END>)
PsiElement(<DOC_DATA>)
```

**Comment by izhangzhihao** [comment]
The release version:

![image](https://user-images.githubusercontent.com/23321756/92714397-d55ad380-f364-11ea-8aed-d386dce283af.png)

The nightly version:

![image](https://user-images.githubusercontent.com/12044174/121648855-87792480-caca-11eb-8fd2-507be0974775.png)

The code highlight are missing

**Comment by izhangzhihao** [comment]
As I see, the `PsiElement(<DOC_DATA>)` just as `PsiComment(<INNER_EOL_DOC_COMMENT>)`, is not structured, I could not analyze unstructured element and highlighting them.

**Comment by izhangzhihao** [comment]
But just for the issue https://github.com/izhangzhihao/intellij-rainbow-brackets/issues/483, the height issue got fixed(if the way we fix it is removing it), yes.

**Comment by vlad20012** [comment]
@izhangzhihao hello!
I see you are not too familiar with Rust. The first screenshot is from the issue, but the second is your attempt to reproduce it. I guess you're doing it a bit wrong, and this is the reason you don't see the highlighting. Let me explain.

In Rust, there is a thing called `rustdoc`. It's a kind of documentation for rust code parts (functions, structs, modules, etc).
```rust
/// rustdoc for the function
fn some_function() {}
```

The language for rustdoc is Markdown (particularly, some extension of [Common Mark](https://spec.commonmark.org/0.29/) Markdown flavor).

```rust
/// rustdoc for the function.
/// This actually a markdown, so I can use `code snippets` or **bold text**
fn some_function() {}
```

Markdown has a feature called `code fence`:

````rust
/// rustdoc for the function.
/// ```
/// this is a code fence
/// ```
fn some_function() {}
````

Then, Rust has a feature called [`doctest`](https://doc.rust-lang.org/rustdoc/documentation-tests.html) (documentation test).
The idea is that you can write a Rust code is code fences in rustdoc, and this code is actually **a test** (!) can be compiled and ran (as a test). 

_Not every_ Markdown code fence in rustdoc is a doctest. If you want to write a doctest, you have to create a cargo package with a library target (using `cargo new` command). Code fences in non-library targets (e.g. in binary) or in Rust files without cargo project are not considered as doctests and are _not highlighted as Rust code by IntelliJ-Rust plugin_.

**Comment by vlad20012** [comment]
@izhangzhihao 
Could you please clarify your issue a little (https://github.com/izhangzhihao/intellij-rainbow-brackets/issues/483)?

**Comment by izhangzhihao** [comment]
Thanks for your tips! 

After I set up the cargo environment, it still does not work, am I missing something?

Cargo.toml:

```
[package]
name = "rust"
version = "0.1.0"
authors = [""]
edition = "2018"

[lib]
name = "test"
crate-type = ["staticlib"]

[dependencies]

```

The sample code under `src/lib.rs`.

```
cargo -V
cargo 1.51.0
```

The project attached as a cargo project by Intellij:

```
IntelliJ IDEA 2021.2 EAP (Ultimate Edition)
Build #IU-212.4037.9, built on June 8, 2021
IntelliJ IDEA EAP User
Expiration date: July 8, 2021
Runtime version: 11.0.11+9-b1481.1 x86_64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
macOS 11.2.3
GC: G1 Young Generation, G1 Old Generation
Memory: 4096M
Cores: 16
Registry: ide.tooltip.initialDelay=900, ide.balloon.shadow.size=0, scala.erase.compiler.process.jdk.once=false
Non-Bundled Plugins: atom-one (1.0), org.nik.presentation-assistant (1.0.9), org.toml.lang (0.2.148.3911-212), kotest-plugin-intellij (1.1.36-IC-2021.1), Pythonid (212.4037.9), org.intellij.scala (2021.2.9), izhangzhihao.rainbow.brackets (6.17), org.rust.lang (0.4.149.3918-212-nightly)
Kotlin: 212-1.4.32-release-IJ4037.9
```



And by the way, I tried the language injection:

![image](https://user-images.githubusercontent.com/12044174/121658187-a9c37000-cad3-11eb-88a4-943c9cfcc0c2.png)

nothing happened...

**Comment by vlad20012** [comment]
> am I missing something?

Yes, `crate-type = ["staticlib"]` line in `Cargo.toml` also disabled doctests. It will work if you just remove it

**Comment by izhangzhihao** [comment]
![image](https://user-images.githubusercontent.com/12044174/121659135-9369e400-cad4-11eb-91cb-97b42ea1241c.png)

**Comment by izhangzhihao** [comment]
I don't know if I understand it right, every line of code is a `PsiElement(<DOC_DATA>)`?

![image](https://user-images.githubusercontent.com/12044174/121659515-f3608a80-cad4-11eb-9504-0f0657a5cf86.png)

**Comment by izhangzhihao** [comment]
I still couldn't know how the internal `DOC_DATA`  highlighted...

**Comment by vlad20012** [comment]
Yes, `PsiElement(<DOC_DATA>)`. Rust highlighting inside it is a [`language injection`](https://www.jetbrains.com/help/idea/using-language-injections.html) (via `RsDoctestLanguageInjector`).

**Comment by izhangzhihao** [comment]
Do you have no idea why the first element of the rust doc will be highlighted wrongly?

**Comment by vlad20012** [comment]
Well, I see some weird highlighting on your [screenshot](https://github.com/intellij-rust/intellij-rust/issues/6100#issuecomment-859396621) (in 3 places), but I have no idea what is it. I mean, I know nothing about your plugin, I don't know what it highlights and how it highlights. If you can tell me more about this, I think I can answer the question :)

**Comment by izhangzhihao** [comment]
By default(the Rust case), we will get matched brace pair [PairedBraceMatcher](https://github.com/izhangzhihao/intellij-rainbow-brackets/blob/2020.3/src/main/kotlin/com/github/izhangzhihao/rainbow/brackets/BracePairs.kt#L25)

And using the default visitor to highlighting the brackets [DefaultRainbowVisitor](https://github.com/izhangzhihao/intellij-rainbow-brackets/blob/2020.3/src/main/kotlin/com/github/izhangzhihao/rainbow/brackets/visitor/DefaultRainbowVisitor.kt#L14)

So the basic logic is I visit all the elements of the document, [and count the brackets level](https://github.com/izhangzhihao/intellij-rainbow-brackets/blob/2020.3/src/main/kotlin/com/github/izhangzhihao/rainbow/brackets/visitor/DefaultRainbowVisitor.kt#L29), after I got the level, [I will hight it](https://github.com/izhangzhihao/intellij-rainbow-brackets/blob/2020.3/src/main/kotlin/com/github/izhangzhihao/rainbow/brackets/visitor/DefaultRainbowVisitor.kt#L36).

**Comment by izhangzhihao** [comment]
I guess what the screenshot says is: the first element of the code snippet and the last element were wrongly recognized as a pair. That could explain why the first element and the last was yellow because the first level of brackets are yellow by default.

**Comment by izhangzhihao** [comment]
After debugging the sample code:

```
//! ```
//! let optional = None;
//! ```
```

I see some keyword  and brackets appeared, might related to some string manipulation 

https://github.com/intellij-rust/intellij-rust/blob/2cdcc5f27d68a9c3d92f035265ce96a1b2e4adcc/src/main/kotlin/org/rust/ide/injected/RsDoctestLanguageInjector.kt#L138-L140

https://github.com/intellij-rust/intellij-rust/blob/2cdcc5f27d68a9c3d92f035265ce96a1b2e4adcc/src/main/kotlin/org/rust/ide/injected/RsDoctestLanguageInjector.kt#L146

https://github.com/intellij-rust/intellij-rust/blob/2cdcc5f27d68a9c3d92f035265ce96a1b2e4adcc/src/main/kotlin/org/rust/ide/injected/RsDoctestLanguageInjector.kt#L149

**Comment by vlad20012** [comment]
Well, yes, the injected code is preprocessed a bit and there are invisible parts. You can see the full text using `Edit Rust Fragment` intention action in alt+enter menu:
![image](https://user-images.githubusercontent.com/3221931/121671274-04020800-cab7-11eb-8edb-5cc97a5caf48.png)

The bug may be related to these invisible braces somehow

**Comment by vlad20012** [comment]
@izhangzhihao I think I found the truth!

I tried to highlight left and right braces by myself. Here is the result:
![image](https://user-images.githubusercontent.com/3221931/121673202-53493800-cab9-11eb-906a-14b31f6c01b9.png)

Haha!

I think the bug is not in your plugin. It's either in our plugin or in the IntelliJ platform.

**Comment by vlad20012** [comment]
The fix looks very trivial, so I fixed this is #7333. But I believe it is a bug in the platform. For example, I guess the same could be reproduced for Java (with your `rainbow-brackets` plugin:
```java
import org.intellij.lang.annotations.Language;

class Scratch {
    public static void main(String[] args) {
        foo("int i = 0;"); // The bug should appear here
    }

    static void foo(
            @Language(
                    value = "JAVA",
                    prefix = "class A { void a(int foo) {",
                    suffix = "} }"
            ) String t) { }
}
```

Or even for Java without any extra plugins:

```Java
import org.intellij.lang.annotations.Language;

static class Scratch {
    public static void main(String[] args) {
        foo("*/"); // the astersk is highlighted wrongly
    }

    static void foo(
            @Language(
                    value = "JAVA",
                    prefix = "class A {\n    /**\n     * @param foo",
                    suffix = "\n    void a(int foo){} }"
            ) String t) {
        t.length();
    }
}
```

**Comment by izhangzhihao** [comment]
So true, Great minimization.
